Signal/src/util/OWSBackupImportJob.m

```diff
@@ -182,118 +201,174 @@ - (void)downloadAndProcessManifest:(OWSBackupJobCompletion)completion
 {
     OWSAssert(completion);
+    OWSAssert(self.backupIO);
 
     if (self.isComplete) {
         return;
     }
 
     DDLogVerbose(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
 
+    NSData *_Nullable manifestDataDecrypted =
+        [self.backupIO decryptDataAsData:manifestDataEncrypted encryptionKey:self.delegate.backupEncryptionKey];
+    if (!manifestDataDecrypted) {
+        OWSProdLogAndFail(@"%@ Could not decrypt manifest.", self.logTag);
+        return completion(NO);
+    }
+
     NSError *error;
     NSDictionary<NSString *, id> *_Nullable json =
-        [NSJSONSerialization JSONObjectWithData:manifestData options:0 error:&error];
+        [NSJSONSerialization JSONObjectWithData:manifestDataDecrypted options:0 error:&error];
     if (![json isKindOfClass:[NSDictionary class]]) {
         OWSProdLogAndFail(@"%@ Could not download manifest.", self.logTag);
         return completion(NO);
     }
 
     DDLogVerbose(@"%@ json: %@", self.logTag, json);
 
-    NSDictionary<NSString *, NSString *> *_Nullable databaseRecordMap = json[kOWSBackup_ManifestKey_DatabaseFiles];
-    NSDictionary<NSString *, NSString *> *_Nullable attachmentRecordMap = json[kOWSBackup_ManifestKey_AttachmentFiles];
-    NSString *_Nullable databaseKeySpecBase64 = json[kOWSBackup_ManifestKey_DatabaseKeySpec];
-    if (!([databaseRecordMap isKindOfClass:[NSDictionary class]] &&
-            [attachmentRecordMap isKindOfClass:[NSDictionary class]] &&
-            [databaseKeySpecBase64 isKindOfClass:[NSString class]])) {
-        OWSProdLogAndFail(@"%@ Invalid manifest.", self.logTag);
-        return completion(NO);
-    }
-    NSData *_Nullable databaseKeySpec = [NSData dataFromBase64String:databaseKeySpecBase64];
-    if (!databaseKeySpec) {
-        OWSProdLogAndFail(@"%@ Invalid manifest databaseKeySpec.", self.logTag);
+    NSArray<OWSBackupImportItem *> *_Nullable databaseItems =
+        [self parseItems:json key:kOWSBackup_ManifestKey_DatabaseFiles];
+    if (!databaseItems) {
         return completion(NO);
     }
-    if (![OWSBackupJob storeDatabaseKeySpec:databaseKeySpec keychainKey:kOWSBackup_ImportDatabaseKeySpec]) {
-        OWSProdLogAndFail(@"%@ Couldn't store databaseKeySpec from manifest.", self.logTag);
+    NSArray<OWSBackupImportItem *> *_Nullable attachmentsItems =
+        [self parseItems:json key:kOWSBackup_ManifestKey_AttachmentFiles];
+    if (!attachmentsItems) {
         return completion(NO);
     }
 
-    self.databaseRecordMap = [databaseRecordMap mutableCopy];
-    self.attachmentRecordMap = [attachmentRecordMap mutableCopy];
+    self.databaseItems = databaseItems;
+    self.attachmentsItems = attachmentsItems;
 
     return completion(YES);
 }
 
-- (void)downloadFilesFromCloud:(NSMutableArray<NSString *> *)recordNames completion:(OWSBackupJobCompletion)completion
+- (nullable NSArray<OWSBackupImportItem *> *)parseItems:(id)json key:(NSString *)key
```

Comment:
One nice thing about using "items" is that all of the (defensive) parsing logic is gathered in a single place.